### PR TITLE
add option to split recorded history files

### PR DESCRIPTION
**Pull request by Zayankovsky**
### Description
This PR introduces several new Copilot options:
1. Split history files by duration. 
2. Limit number of history files to be uploaded per session. 
3. Limit total size of history files to be uploaded per session.

### CODE REVIEW

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/options/CopilotOptions.kt [hunk 1]
```diff
@@ -74,6 +101,39 @@ class CopilotOptions private constructor(
             this.shouldSendHistoryOnlyWithFeedback = flag
         }
 
+        /**
+         * Limits duration of a history recording. If a trip takes longer than this value,
+         * the history file will be split into several parts. Default is [Long.MAX_VALUE]
+         */
+        fun maxHistoryFileLengthMillis(millis: Long): Builder = apply {
```
Comment (dzinad) [inline libnavigation-base/src/main/java/com/mapbox/navigation/base/options/CopilotOptions.kt:108]: current.txt needs to be updated.

File: libnavigation-copilot/src/main/java/com/mapbox/navigation/copilot/MapboxCopilotImpl.kt [hunk 2]
```diff
@@ -360,9 +394,30 @@ internal class MapboxCopilotImpl(
...
     }
 
-    private suspend fun pushHistoryFile(historyFilePath: String, drive: CopilotMetadata) {
+    private fun limitTotalHistoryFilesSize(
+        historyFiles: MutableMap<File, CopilotMetadata>,
+        minFilesCount: Int,
+    ) {
+        var total = historyFiles.keys.sumOf { HistoryAttachmentsUtils.size(it) }
+        while (total > maxTotalHistoryFilesSizePerSession && historyFiles.size > minFilesCount) {
+            val firstFile = historyFiles.keys.first()
```
Comment (dzinad) [inline libnavigation-copilot/src/main/java/com/mapbox/navigation/copilot/MapboxCopilotImpl.kt:403]: Is the order preserved here?

File: libnavigation-copilot/src/main/java/com/mapbox/navigation/copilot/MapboxCopilotImpl.kt [hunk 3]
```diff
@@ -316,17 +346,21 @@ internal class MapboxCopilotImpl(
...
+                limitTotalHistoryFilesSize(historyFilesCopy, minFilesCount = 1)
+                pushHistoryFiles(historyFilesCopy)
             }
         } else {
             cancelRecordingHistory()
         }
     }
 
     private fun cancelRecordingHistory() {
+        for (historyFile in historyFiles.keys) {
```
Comment (dzinad) [inline libnavigation-copilot/src/main/java/com/mapbox/navigation/copilot/MapboxCopilotImpl.kt:361]: Should you also clear the map here?

File: libnavigation-copilot/src/main/java/com/mapbox/navigation/copilot/MapboxCopilotImpl.kt (hunk 2 above)
Comment (Zayankovsky) [reply libnavigation-copilot/src/main/java/com/mapbox/navigation/copilot/MapboxCopilotImpl.kt:403]: Yes, `mutableMapOf` creates a map that uses insertion-oder for iteration. It is actually a `LinkedHashMap`. Also `toMutableMap` preserves the iteration order.

File: libnavigation-copilot/src/main/java/com/mapbox/navigation/copilot/MapboxCopilotImpl.kt (hunk 3 above)
Comment (Zayankovsky) [reply libnavigation-copilot/src/main/java/com/mapbox/navigation/copilot/MapboxCopilotImpl.kt:361]: It is cleared in `stopRecording`.

**Review by dzinad — APPROVED** [review]

**Review by Guardiola31337 — APPROVED** [review]
Looks good to me 👍 

Thanks for the additions 🙇‍♂️

**Pull request merged**